File: src/padmap/assign.py
"""Press-and-hold controller assignment.

For pads that are indistinguishable by every static attribute, a human
pressing a button is the only available source of identity. This module turns
that press into a player ordering.

A *held* button is required rather than a single rising edge. During earlier
testing an empty adapter port registered a stray press that a first-edge
scheme would have accepted, silently burning a player slot. A transient cannot
hold; a human cannot tell the difference.
"""

from __future__ import annotations

import logging
import select
import time
from dataclasses import dataclass
from typing import Any, Callable, Iterable

import evdev
from evdev import ecodes

from . import hidraw
from .devices import Pad, open_device

log = logging.getLogger("padmap.assign")

HOLD_SECONDS = 0.25
BTN_FIRST = 0x100  # ignore the keyboard range on combo devices


@dataclass
class Assignment:
    player: int  # 1-based
    pad: Pad
    button: int  # the BTN_* code that claimed the slot


# (pad, held_fraction) -> None. Lets a UI draw a fill/progress ring.
ProgressFn = Callable[[Pad, float], None]
# (assignment) -> None, fired the moment a slot is claimed.
ClaimFn = Callable[[Assignment], None]
# (pad, button code, value) -> None, for button activity on an already-claimed
# pad. A UI can build "hold again to continue" out of this.
ClaimedEventFn = Callable[[Pad, int, int], None]
# (pad, event) -> None, for every event read from any pad. Calibration needs
# the EV_ABS traffic that claim detection ignores, and the pads are grabbed
# during a session so nothing else can read them.
RawEventFn = Callable[[Pad, "evdev.InputEvent"], None]


class Assigner:
    """Watches a set of pads and yields player order from held presses."""

    def __init__(
        self,
        pads: Iterable[Pad],
        grab: bool = True,
        hold_seconds: float = HOLD_SECONDS,
    ) -> None:
        self.pads = list(pads)
        self.hold_seconds = hold_seconds
        self._grab = grab
        # `Any`, not `evdev.InputDevice`, because a hidraw pad is not one --
        # `hidraw.Source` wears the shape of an InputDevice and implements only
        # the twelve members this class and the republisher actually use. Saying
        # InputDevice here was a lie mypy caught and nothing else would have:
        # the two types diverge exactly where force feedback is, and calling a
        # member Source does not have raises AttributeError, which is not an
        # OSError and so passes through every guard between here and the
        # selector loop. Same annotation virtual.create already uses, for the
        # same reason.
        self._devices: dict[int, Any] = {}
        self._pad_by_fd: dict[int, Pad] = {}
        # fd -> (button code, monotonic time it went down)
        self._holding: dict[int, tuple[int, float]] = {}
        self._claimed: set[int] = set()
        self.assignments: list[Assignment] = []
        # Set by a caller that wants button activity from already-claimed pads.
        self.on_claimed_event: ClaimedEventFn | None = None
        # Set by a caller that needs the raw stream, e.g. axis calibration.
        self.on_raw_event: RawEventFn | None = None
        # Pads that could not be grabbed exclusively; their presses reach
        # other applications as well as us.
        self.grab_failures: list[Pad] = []

    def __enter__(self) -> "Assigner":
        for pad in self.pads:
            # The same source the republisher picks, not always the evdev
            # node. A pad that has to be read over hidraw has to be read that
            # way *here too*: the assigner is what the setup screen and the
            # mapping wizard listen through, so reading a different device
            # from the one being republished means the screen that maps a
            # controller cannot see the controller it is mapping.
            device = hidraw.open_source(pad) or open_device(pad)
            if self._grab:
                # Keep setup presses out of whatever else has focus.
                try:
                    device.grab()
                except OSError as exc:
                    # Worth saying out loud rather than swallowing: a failed
                    # grab means every press during setup ALSO reaches the
                    # front-end, so the button that claims a slot doubles as a
                    # UI keypress. Usually means something else already holds
                    # an exclusive grab on the device.
                    log.warning("could not grab %s (%s): presses will leak "
                                "through to other applications",
                                pad.event, exc)
                    self.grab_failures.append(pad)
            self._devices[device.fd] = device
            self._pad_by_fd[device.fd] = pad
        self._drain()
        return self

    def __exit__(self, *_exc) -> None:
        self.close()

    def close(self) -> None:
        for device in self._devices.values():
            if self._grab:
                try:
                    device.ungrab()
                except OSError:
                    pass
            device.close()
        self._devices.clear()

    def _drain(self) -> None:
        """Discard anything queued before we started listening."""
        for device in self._devices.values():
            try:
                while device.read_one() is not None:
                    pass
            except OSError:
                pass

    # -- event-loop integration -------------------------------------------
    #
    # `run` below owns a select loop, which cannot coexist with a GUI toolkit's
    # own loop. These three let an external loop drive the same state machine:
    # watch `fds` for readability, call `handle_readable` when one fires, and
    # call `tick` on a short timer. The timer is not optional -- a held button
    # emits no further events, so a hold completing is only ever noticed by
    # looking at the clock.

    @property
    def fds(self) -> list[int]:
        """Device file descriptors to watch for readability."""
        return list(self._devices)

    def handle_readable(self, fd: int) -> None:
        """Consume pending events on a descriptor reported readable."""
        self._consume(fd)

    def device_for(self, pad: Pad) -> evdev.InputDevice | None:
        """The open handle for a pad, or None if it is not in this session.

        Exposed so calibration can read absinfo without opening a second
        handle: the pads are grabbed here, and a second reader would compete
        for the same events.
        """
        for fd, candidate in self._pad_by_fd.items():
            if candidate.path == pad.path:
                return self._devices.get(fd)
        return None

    def tick(
        self,
        on_progress: ProgressFn | None = None,
        on_claim: ClaimFn | None = None,
    ) -> None:
        """Advance hold timers. Call roughly every 20ms."""
        self._check_holds(on_progress, on_claim)

    def run(
        self,
        wanted: int | None = None,
        timeout: float | None = None,
        on_progress: ProgressFn | None = None,
        on_claim: ClaimFn | None = None,
    ) -> list[Assignment]:
        """Collect assignments until `wanted` slots are filled or time runs out.

        `wanted=None` means "until every discovered pad is claimed", which is
        rarely what you want with multi-port adapters -- most of their ports
        have nothing plugged in and will never press anything.
        """
        target = wanted if wanted is not None else len(self.pads)
        deadline = None if timeout is None else time.monotonic() + timeout

        while len(self.assignments) < target:
            now = time.monotonic()
            if deadline is not None and now >= deadline:
                break

            # Wake often enough to notice a hold completing, since a held
            # button generates no further events on its own.
            wait = 0.02
            if deadline is not None:
                wait = min(wait, max(0.0, deadline - now))

            readable, _, _ = select.select(self.fds, [], [], wait)
            for fd in readable:
                self.handle_readable(fd)

            self.tick(on_progress, on_claim)

        return self.assignments

    def reset(self) -> None:
        """Drop every claim and start over.

        Also drains queued events, so a button still held from the previous
        round cannot immediately re-claim a slot.
        """
        self.assignments.clear()
        self._claimed.clear()
        self._holding.clear()
        self._drain()

    def _consume(self, fd: int) -> None:
        device = self._devices.get(fd)
        if device is None:
            return
        try:
            events = list(device.read())
        except (OSError, BlockingIOError):
            return

        if self.on_raw_event is not None:
            pad = self._pad_by_fd[fd]
            for event in events:
                self.on_raw_event(pad, event)

        if fd in self._claimed:
            # Claimed pads still have to be drained. They are grabbed, so
            # nothing else will empty their buffer, and an unread descriptor
            # stays permanently readable -- which would make any select-based
            # loop spin at 100% CPU. Forwarding the events also gives callers
            # a way to build a confirm gesture on an assigned pad.
            if self.on_claimed_event is not None:
                pad = self._pad_by_fd[fd]
                for event in events:
                    if event.type == ecodes.EV_KEY and event.code >= BTN_FIRST:
                        self.on_claimed_event(pad, event.code, event.value)
            return

        for event in events:
            if event.type != ecodes.EV_KEY or event.code < BTN_FIRST:
                continue
            if event.value == 1:
                # Track only the first button down; a second button pressed
                # during a hold should not restart the timer.
                if fd not in self._holding:
                    self._holding[fd] = (event.code, time.monotonic())
            elif event.value == 0:
                held = self._holding.get(fd)
                if held and held[0] == event.code:
                    # Released early: a tap, or a transient. Not a claim.
                    del self._holding[fd]

    def _check_holds(
        self, on_progress: ProgressFn | None, on_claim: ClaimFn | None
    ) -> None:
        now = time.monotonic()
        for fd, (code, started) in list(self._holding.items()):
            if fd in self._claimed:
                continue
            elapsed = now - started
            if elapsed < self.hold_seconds:
                if on_progress:
                    on_progress(self._pad_by_fd[fd], elapsed / self.hold_seconds)
                continue

            pad = self._pad_by_fd[fd]
            assignment = Assignment(
                player=len(self.assignments) + 1, pad=pad, button=code
            )
            self.assignments.append(assignment)
            self._claimed.add(fd)
            del self._holding[fd]
            if on_progress:
                on_progress(pad, 1.0)
            if on_claim:
                on_claim(assignment)

File: rust/crates/padmap-core/src/assign.rs
//! Press-and-hold controller assignment: a held button claims a slot, not a single press.

use std::collections::BTreeMap;

/// How long a button must be down to claim a slot, for whoever does not ask.
pub const HOLD_SECONDS: f64 = 0.25;

/// What a hold can be set to: nought is a press, and a minute is not a hold.
pub const HOLD_RANGE: std::ops::RangeInclusive<f64> = 0.05..=10.0;

/// The hold length asked for, or [`HOLD_SECONDS`] where none was or it is
/// outside [`HOLD_RANGE`]. A comfort setting is not worth refusing over, so
/// there is no error here and NaN falls through like any other non-length.
pub fn hold_or_default(seconds: Option<f64>) -> f64 {
    seconds
        .filter(|seconds| HOLD_RANGE.contains(seconds))
        .unwrap_or(HOLD_SECONDS)
}

/// The same, from text: what an environment variable carries.
pub fn hold_from(text: Option<&str>) -> f64 {
    hold_or_default(text.and_then(|text| text.trim().parse::<f64>().ok()))
}

/// Ignore the keyboard range on combo devices (first valid button code).
pub const BTN_FIRST: u16 = 0x100;

/// One pad's claim on a player slot.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Assignment {
    /// 1-based.
    pub player: u32,
    /// Index into the caller's pad list.
    pub pad: usize,
    /// The button code that claimed it.
    pub button: u16,
}

/// What a tick produced.
#[derive(Debug, Clone, PartialEq)]
pub struct Tick {
    /// Pads with a hold in flight, and how far through it they are (0.0..1.0).
    pub progress: Vec<(usize, f64)>,
    /// Slots claimed by this tick.
    pub claimed: Vec<Assignment>,
}

/// Watches a set of pads and yields player order from held presses.
#[derive(Debug, Clone)]
pub struct Assigner {
    hold_seconds: f64,
    /// pad index -> (button code, when it went down).
    holding: BTreeMap<usize, (u16, f64)>,
    claimed: Vec<usize>,
    assignments: Vec<Assignment>,
}

impl Default for Assigner {
    fn default() -> Self {
        Assigner::new(HOLD_SECONDS)
    }
}

impl Assigner {
    pub fn new(hold_seconds: f64) -> Self {
        Assigner {
            hold_seconds,
            holding: BTreeMap::new(),
            claimed: Vec::new(),
            assignments: Vec::new(),
        }
    }

    pub fn assignments(&self) -> &[Assignment] {
        &self.assignments
    }

    /// How long a hold has to run here to claim.
    pub fn hold_seconds(&self) -> f64 {
        self.hold_seconds
    }

    /// Set how long a hold has to run, dropping every hold in flight: a press
    /// that became a claim because the number moved under it is exactly the
    /// accident a longer hold exists to prevent.
    pub fn set_hold_seconds(&mut self, hold_seconds: f64) {
        self.hold_seconds = hold_seconds;
        self.holding.clear();
    }

    /// Whether this pad has already claimed a slot.
    pub fn is_claimed(&self, pad: usize) -> bool {
        self.claimed.contains(&pad)
    }

    /// Offer one event. Only EV_KEY codes above BTN_FIRST count.
    pub fn feed(&mut self, pad: usize, kind: u16, code: u16, value: i32, now: f64) {
        if kind != crate::capture::EV_KEY || code < BTN_FIRST {
            return;
        }
        if value == 1 {
            self.holding.entry(pad).or_insert((code, now));
        } else if value == 0 {
            // Only the button that started the hold: a thumb lifting off B must not
            // cancel a hold running on A.
            if self.holding.get(&pad).map(|(held, _)| *held) == Some(code) {
                self.holding.remove(&pad);
            }
        }
    }

    /// Advance the hold timers. Must be called on a timer (events alone cannot detect completion).
    pub fn tick(&mut self, now: f64) -> Tick {
        let mut out = Tick {
            progress: Vec::new(),
            claimed: Vec::new(),
        };
        let pending: Vec<(usize, (u16, f64))> = self
            .holding
            .iter()
            .map(|(pad, held)| (*pad, *held))
            .collect();

        for (pad, (code, started)) in pending {
            if self.is_claimed(pad) {
                continue;
            }
            let elapsed = now - started;
            if elapsed < self.hold_seconds {
                out.progress
                    .push((pad, (elapsed / self.hold_seconds).clamp(0.0, 1.0)));
                continue;
            }
            let assignment = Assignment {
                player: self.assignments.len() as u32 + 1,
                pad,
                button: code,
            };
            self.assignments.push(assignment);
            self.claimed.push(pad);
            self.holding.remove(&pad);
            out.progress.push((pad, 1.0));
            out.claimed.push(assignment);
        }
        out
    }

    /// Drop every claim. Caller must discard queued pad events.
    pub fn reset(&mut self) {
        self.assignments.clear();
        self.claimed.clear();
        self.holding.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::capture::{EV_ABS, EV_KEY};

    const A: u16 = 0x130;
    const B: u16 = 0x131;

    #[test]
    fn a_held_button_claims_the_first_free_slot() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assert!(assigner.tick(0.1).claimed.is_empty(), "too soon");
        let tick = assigner.tick(HOLD_SECONDS + 0.01);
        assert_eq!(tick.claimed.len(), 1);
        assert_eq!(
            tick.claimed[0],
            Assignment {
                player: 1,
                pad: 0,
                button: A
            }
        );
    }

    #[test]
    fn a_length_is_read_from_text_or_is_the_default() {
        assert_eq!(hold_from(Some("1.5")), 1.5);
        assert_eq!(hold_from(Some("  0.05  ")), 0.05, "the ends are inside");
        assert_eq!(hold_from(Some("10")), 10.0);
        // Nothing here is worth refusing to start over.
        for asked in ["", "soon", "0", "-2", "600", "nan", "inf", "1,5"] {
            assert_eq!(hold_from(Some(asked)), HOLD_SECONDS, "{asked:?}");
        }
        assert_eq!(hold_from(None), HOLD_SECONDS);
        assert_eq!(hold_or_default(Some(f64::NAN)), HOLD_SECONDS);
    }

    #[test]
    fn a_longer_hold_takes_longer_and_fills_at_its_own_pace() {
        let mut assigner = Assigner::new(1.5);
        assert_eq!(assigner.hold_seconds(), 1.5);
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assert!(assigner.tick(0.3).claimed.is_empty(), "0.3s is not a claim");
        assert!(assigner.tick(1.4).claimed.is_empty(), "nor is 1.4s");
        // Progress fills over whatever length is set, not over the default.
        let part = assigner.tick(0.75).progress[0].1;
        assert!((part - 0.5).abs() < 0.01, "progress at half was {part}");
        assert_eq!(assigner.tick(1.51).claimed.len(), 1);
    }

    #[test]
    fn changing_the_length_drops_the_hold_running_under_it() {
        let mut assigner = Assigner::new(1.5);
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.tick(1.0);
        // A press that became a claim because the number moved is the accident.
        assigner.set_hold_seconds(0.25);
        assert!(
            assigner.tick(1.1).claimed.is_empty(),
            "the hold in flight claimed on the new length"
        );
        assert!(assigner.tick(100.0).claimed.is_empty(), "and never does");
        // A fresh press measures the new length.
        assigner.feed(0, EV_KEY, A, 1, 2.0);
        assert_eq!(assigner.tick(2.26).claimed.len(), 1);
    }

    #[test]
    fn a_tap_claims_nothing() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.feed(0, EV_KEY, A, 0, 0.05);
        assert!(assigner.tick(1.0).claimed.is_empty());
        assert!(assigner.assignments().is_empty());
    }

    #[test]
    fn a_second_button_during_a_hold_does_not_restart_the_timer() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.feed(0, EV_KEY, B, 1, 0.2);
        assert_eq!(assigner.tick(HOLD_SECONDS + 0.01).claimed.len(), 1);
    }

    #[test]
    fn releasing_a_different_button_does_not_cancel_the_hold() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.feed(0, EV_KEY, B, 1, 0.05);
        assigner.feed(0, EV_KEY, B, 0, 0.10);
        assert_eq!(assigner.tick(HOLD_SECONDS + 0.01).claimed.len(), 1);
    }

    #[test]
    fn players_are_numbered_in_the_order_the_pads_claimed() {
        let mut assigner = Assigner::default();
        assigner.feed(2, EV_KEY, A, 1, 0.0);
        assigner.tick(HOLD_SECONDS + 0.01);
        assigner.feed(0, EV_KEY, A, 1, 1.0);
        assigner.tick(1.0 + HOLD_SECONDS + 0.01);
        let players: Vec<(u32, usize)> = assigner
            .assignments()
            .iter()
            .map(|a| (a.player, a.pad))
            .collect();
        assert_eq!(players, [(1, 2), (2, 0)]);
    }

    #[test]
    fn one_pad_cannot_claim_twice() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.tick(HOLD_SECONDS + 0.01);
        assigner.feed(0, EV_KEY, B, 1, 1.0);
        assert!(assigner.tick(1.0 + HOLD_SECONDS + 0.01).claimed.is_empty());
        assert_eq!(assigner.assignments().len(), 1);
    }

    #[test]
    fn a_keyboard_code_on_a_combo_adapter_cannot_claim_a_slot() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, 0x1e, 1, 0.0);
        assert!(assigner.tick(1.0).claimed.is_empty());
    }

    #[test]
    fn axis_noise_is_not_a_press() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_ABS, 0x00, 255, 0.0);
        assert!(assigner.tick(1.0).claimed.is_empty());
    }

    #[test]
    fn autorepeat_does_not_start_a_second_hold() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        for step in 1..10 {
            assigner.feed(0, EV_KEY, A, 2, f64::from(step) * 0.01);
        }
        assert_eq!(assigner.tick(HOLD_SECONDS + 0.01).claimed.len(), 1);
    }

    #[test]
    fn progress_climbs_to_one_and_is_reported_for_the_claiming_tick() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        let early = assigner.tick(HOLD_SECONDS / 2.0);
        assert!((early.progress[0].1 - 0.5).abs() < 1e-9);
        let done = assigner.tick(HOLD_SECONDS + 0.01);
        assert_eq!(done.progress, [(0, 1.0)]);
    }

    #[test]
    fn progress_never_exceeds_one() {
        let mut assigner = Assigner::new(1.0);
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        for (_, fraction) in assigner.tick(0.9).progress {
            assert!((0.0..=1.0).contains(&fraction));
        }
    }

    #[test]
    fn reset_drops_every_claim() {
        let mut assigner = Assigner::default();
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        assigner.tick(HOLD_SECONDS + 0.01);
        assigner.reset();
        assert!(assigner.assignments().is_empty());
        assert!(!assigner.is_claimed(0));
        assigner.feed(0, EV_KEY, A, 1, 2.0);
        assert_eq!(assigner.tick(2.0 + HOLD_SECONDS + 0.01).claimed.len(), 1);
    }

    #[test]
    fn two_pads_holding_at_once_both_claim_in_pad_order() {
        let mut assigner = Assigner::default();
        assigner.feed(1, EV_KEY, A, 1, 0.0);
        assigner.feed(0, EV_KEY, A, 1, 0.0);
        let tick = assigner.tick(HOLD_SECONDS + 0.01);
        assert_eq!(
            tick.claimed
                .iter()
                .map(|a| (a.player, a.pad))
                .collect::<Vec<_>>(),
            [(1, 0), (2, 1)]
        );
    }
}
